### ISSUE TITLE
[TS React Apollo] Variables should be optional for Query components

### ISSUE THREAD

**Issue by cecchi** [dotansimha/graphql-code-generator#1897]
**Describe the bug**

The generated types for React Apollo `<Query>` components explicitly override the "variables" prop and makes it required:

```ts
export type NavHierarchyComponentProps = Omit<
  Omit<ReactApollo.QueryProps<NavHierarchyQuery, NavHierarchyQueryVariables>, 'query'>,
  'variables'
> & { variables: NavHierarchyQueryVariables }
```

It looks as though this was done intentionally in https://github.com/dotansimha/graphql-code-generator/pull/1687, and then undone for mutations in https://github.com/dotansimha/graphql-code-generator/issues/1741, so I'm probably missing some context as to why that was desirable. 

Our use-case is that we use the `skip` prop to conditionally execute the query only once the variables are available. Something like this:

```jsx
<NavHierarchyComponent 
  variables={someContext ? { foo: someContext.bar } : undefined} 
  skip={someContext === undefined} 
/>
```

This fails with the current typings since it expects `variables` to be explicitly non-optional.

**Expected behavior**

Either use the built-in React Apollo types for `variables`, or add some kind of clever typing that allows the `variables` to be optional if the `skip` prop is provided. I believe simply using the typings from `react-apollo` would solve my use-case, but I'm sure there is a reason it was overridden in the first place.

**Edit:** It looks as though this was done since the `react-apollo` typings were wrapping the variables in `Partial`, meaning the types were not as restrictive as they could be (https://github.com/dotansimha/graphql-code-generator/issues/1626). 

In my opinion it's best to lean on the Apollo typings as much as possible, so I would suggest removing the override (`Omit` + `& { variables ... }`) altogether. If they wrap the types in Partial (incorrectly?), that should be dealt with on their end. Anyway, as far as I can tell, this is no longer the case. These are the typings from the latest version of `react-apollo`:

```tsx
export interface QueryOpts<TGraphQLVariables = OperationVariables> {
    variables?: TGraphQLVariables;
}
```

...Which looks correct to me. Variables can be omitted altogether, but if they are present they must exactly match the typings.

**Comment by ardatan** [comment]
If at least one of the variables is required, `variables` property would be required; because this means that query is unable to work without variables. So, codegen generates the component of that query with a required `variables` property.

**Comment by cecchi** [comment]
Thanks @ardatan, the issue is related to the `skip` prop which tells `react-apollo` to not execute the query at all. In that case it's valid to not pass variables.

**Comment by dotansimha** [comment]
@ardatan I think the variables should be the same as they declared in React-Apollo. if they are optional there, let's do the same

**Comment by dotansimha** [comment]
I discussed it with @ardatan , we should allow variables to be `VarType | null` and it will make it easier to use, but still force `variables` to be there when it's needed.
What do you think @cecchi ?

**Comment by cecchi** [comment]
@dotansimha @ardatan, sure, that works -- I'm not positive what assumptions `react-apollo` makes under the hood but if `null` is equivalent to `undefined` then that suits my use-case.

**Comment by ardatan** [comment]
I created a PR to make variables optional if `skip: true` is provided instead of adding `| null`; because it didn't look good to me.
@dotansimha What do you say?

**Comment by dotansimha** [comment]
Available in 1.2.1
